### PR TITLE
Fix test_scheduling

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

Test scheduling was flakey due to timing issues with working startup time, resource reports, and sleeping in the test case.

This fixes it by using semaphores instead of relying on timing.

Note: We need the memory resource, because it's the only one that satisfies the requirements of
* Being part of the critical resource utilization calculation
* Being manually settable (and not auto-filled like object store memory)
* Not borrow-able by `ray.get`

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_scheduling.py [hunk 1]
```diff
@@ -60,35 +61,38 @@ def f():
...
-    for node_ip in node_resources:
-        print("Warming worker pool on ", node_ip)
-        ray.get([get_node.remote() for _ in range(num_cpus)])
+    block_task = Semaphore.remote(0)
+    block_driver = Semaphore.remote(0)
+
+    # Add the memory resource because the cpu will be released in the ray.get
+    @ray.remote(num_cpus=1, memory=1)
+    def get_node():
+        # Sleep to avoid lease reuse.
```
Comment (ijrsvt) [inline python/ray/tests/test_scheduling.py:79]:
```suggestion
```

File: python/ray/tests/test_scheduling.py [hunk 2]
```diff
@@ -60,35 +61,37 @@ def f():
...
 
-    # Warm the worker pool in order to isolate the test to only test the
-    # scheduling policy.
-    node_resources = [
-        resource for resource in ray.cluster_resources() if "node:" in resource
-    ]
-    for node_ip in node_resources:
-        print("Warming worker pool on ", node_ip)
-        ray.get([get_node.remote() for _ in range(num_cpus)])
+    block_task = Semaphore.remote(0)
```
Comment (ijrsvt) [inline python/ray/tests/test_scheduling.py:73]:
```suggestion
	# `block_task` ensures that scheduled tasks do not return until all are running.
    block_task = Semaphore.remote(0)
    # `block_driver` ensures that the driver does not allow tasks to continue until all are running.
```

**Review by ijrsvt — APPROVED** [review]
This is a great modification to the test and I really like how it works. I tried to add two comments to make it more clear to someone skimming the file.  IDK if they help at all  ¯\\\_(ツ)_/¯

**Comment by wuisawesome** [comment]
Test failures look unrelated. the flakey test is passing now. merging

**Comment by ericl** [comment]
Great fix!

**Pull request merged**